catkin_tools/verbs/catkin_build/job.py

```diff
@@ -107,24 +107,38 @@ def next(self):
 
 
 def create_env_file(package, context):
+    # Exporting _CATKIN_SETUP_DIR supports DESTDIR, where the absolute location of the installspace
+    # at build time does not match the intended final location of it.
+    source_snippet = '_CATKIN_SETUP_DIR={space_path} . "{source_path}"'
     sources = []
-    source_snippet = '. "{source_path}"'
     # If installing to isolated folders or not installing, but devel spaces are not merged
     if (context.install and context.isolate_install) or (not context.install and context.isolate_devel):
         # Source each package's install or devel space
         space = context.install_space_abs if context.install else context.devel_space_abs
-        # Get the recursive dependcies
+        # Get the recursive dependencies
         depends = get_cached_recursive_build_depends_in_workspace(package, context.packages)
         # For each dep add a line to source its setup file
         for dep_pth, dep in depends:
-            source_path = os.path.join(space, dep.name, 'setup.sh')
-            sources.append(source_snippet.format(source_path=source_path))
+            space_path = os.path.join(space, dep.name)
+            source_path = os.path.join(space_path, 'setup.sh')
+            sources.append(source_snippet.format(space_path=space_path, source_path=source_path))
     else:
         # Just source common install or devel space
-        source_path = os.path.join(
-            context.install_space_abs if context.install else context.devel_space_abs,
-            'setup.sh')
-        sources = [source_snippet.format(source_path=source_path)] if os.path.exists(source_path) else []
+        if context.install:
+            if context.destdir:
+                # Intentionally not using os.path.join in this instance, as it's expected that
+                # when DESTDIR is specified, the install space is an absolute path, so a straight
+                # concatenation is more suitable than os.path.join.
+                space_path = context.destdir + context.install_space_abs
```

Comment:
A better explanation (in my opinion) is that since the install space is absolute (and it always is, not just when used in conjunction with `DESTDIR`), `os.path.join` will not return the concatenation by rather just the install space. It's not that this is more suitable than `os.path.join` (which to me implies `os.path.join` would also work) but instead that `os.path.join` doesn't work at all and therefore this is necessary. E.g.:

```
>>> import os
>>> a = '/tmp/build/'
>>> b = '/usr'
>>> os.path.join(a, b)
'/usr'
>>> a = '/tmp/build'
>>> os.path.join(a, b)
'/usr'
```